File: Assets/script/network/NetworkStream.cs
﻿using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

unsafe public class NetworkStream {

    byte[] send_buff = null;
    byte[] read_buff = null;
    int send_position = 0;
    int send_end = 0;
    int read_position = 0;
    int read_end = 0;
    public int ReadPosition
    {
        get { return read_position; }
    }
    public int ReadEnd
    {
        get { return read_end; }
    }
    public int SendPosition
    {
        get { return send_position; }
    }
    public int SendEnd
    {
        get { return send_end; }
    }
    public NetworkStream(int buff_size = 0)
    {
        if (buff_size > 0)
        {
            send_buff = new byte[buff_size];
            read_buff = new byte[buff_size];
        }
    }
    public void SetReadBuffer(byte[] buffer, int data_size)
    {
        read_buff = buffer;
        read_position = 0;
        read_end = data_size;
    }
    public void SetSendBuffer(byte[] buffer, int data_size)
    {
        send_buff = buffer;
        send_position = 0;
        send_end = data_size;
    }
    static void CopyArray(byte* src, byte* dst, int count)
    {
        if (null == src || null == dst || count<=0)
        {
            return;
        }
        if(count>=4)
        {
            for (int n = 0; n < count / 4; n++)
            {
                *((int*)dst) = *((int*)src);
                dst += 4;
                src += 4;
            }
        }
        for (int n = 0; n < count % 4; n++)
        {
            *dst = *src;
            dst++;
            src++;
        }
    }
    //////////////////////////////////////////////////////////////
    //read data
    public byte ReadByte()
    {
        byte reslut = 0;
        ReadData(&reslut, sizeof(byte));
        return reslut;
    }
    public short ReadShort()
    {
        short reslut = 0;
        ReadData(&reslut, sizeof(short));
        return reslut;
    }
    public int ReadInt()
    {
        int reslut = 0;
        ReadData(&reslut, sizeof(int));
        return reslut;
    }
    public float ReadFloat()
    {
        float reslut = 0;
        ReadData(&reslut, sizeof(float));
        return reslut;
    }
    public long ReadLong()
    {
        long reslut = 0;
        ReadData(&reslut, sizeof(long));
        return reslut;
    }

    public ushort ReadUShort()
    {
        ushort reslut = 0;
        ReadData(&reslut, sizeof(ushort));
        return reslut;
    }
    public uint ReadUInt()
    {
        uint reslut = 0;
        ReadData(&reslut, sizeof(uint));
        return reslut;
    }
    public ulong ReadULong()
    {
        ulong reslut = 0;
        ReadData(&reslut, sizeof(ulong));
        return reslut;
    }
    public string ReadString()
    {
        int len = ReadInt();
        if(len>0)
        {
            fixed(byte* data= &read_buff[read_position])
            {
                read_position += len;
                return new string((char*)data,0,len/2);
            }
            
        }
        return string.Empty;
    }
    public void ReadData(byte[] data,int count)
    {
        fixed (byte* b_data = data)
        {
            ReadData(b_data, count);
        }
    }
    public void ReadData(void* data,int count)
    {
        if(read_buff==null)
        {
            throw new Exception("read_buff==null");
        }
        if(read_position + count > read_end)
        {
            throw new Exception("read_position+count>read_end");
        }
        
        fixed (byte* pSrc = &read_buff[read_position])
        {
            CopyArray(pSrc, (byte*)data, count);
        }
        read_position += count;
    }

    //////////////////////////////////////////////////////////////
    //write data
    public void WriteByte(byte data)
    {
        WriteData(&data, sizeof(byte));
    }
    public void WriteShort(short data)
    {
        WriteData(&data, sizeof(short));
    }
    public void WriteInt(int data)
    {
        WriteData(&data, sizeof(int));
    }
    public void WriteFloat(float data)
    {
        WriteData(&data, sizeof(float));
    }
    public void WriteLong(long data)
    {
        WriteData(&data, sizeof(long));
    }

    public void WriteUShort(ushort data)
    {
        WriteData(&data, sizeof(ushort));
    }
    public void WriteUInt(uint data)
    {
        WriteData(&data, sizeof(uint));
    }
    public void WriteULong(ulong data)
    {
        WriteData(&data, sizeof(ulong));
    }
    public void WriteString(string data)
    {
        if(data.Length>0)
        {
            WriteInt(data.Length*2);
            fixed(char* str=data)
            {
                WriteData((byte*)str, data.Length*2);
            }
            
        }
    }
    public void WriteData(byte[] data,int count)
    {
        fixed(byte* b_data=data)
        {
            WriteData(b_data, count);
        }
       
    }
  
    void WriteData(void* data,int count)
    {
        if(send_buff==null)
        {
            throw new System.Exception("send_buff==null");
        }
        if(send_end + count > send_buff.Length)
        {
            throw new System.Exception("send_end+count>send_buff.Length");
        }
        if(count<=0)
        {
            throw new System.Exception("count<=0");
        }
        fixed (byte* pDst = &send_buff[send_end])
        {
            CopyArray((byte*)data, pDst, count);
        }
        send_end += count;
    }
    
}

File: server/src/common/network/NetworkStream.cpp
#include "NetworkStream.h"
#include <string>
NetworkStream::NetworkStream():
	write_buff(nullptr)
	,read_buff(nullptr)
	,write_position(nullptr)
	,write_end(nullptr)
	,read_buff_end(nullptr)
	,write_buff_end(nullptr)
	,read_position(nullptr)
	,read_end(nullptr)
	,send_position(nullptr)
	,send_end(nullptr)
	,write_buff_len(0)
	,read_buff_len(0)
{
}

NetworkStream::~NetworkStream()
{
}
void NetworkStream::SetReadBuffer(char* buffer, int data_size,int buffer_len)
{
	read_buff = buffer;
	read_position = buffer;
	read_end = &buffer[data_size];
	read_buff_len = buffer_len;
	read_buff_end = &buffer[buffer_len - 1];
	memset(buffer, 0, read_end - read_position);
}
void NetworkStream::SetWriteBuffer(char* buffer, int data_size,int buffer_len)
{
	
	write_buff = buffer;
	write_position = buffer;
	write_end = &buffer[data_size];
	write_buff_len = buffer_len;
	send_position = nullptr;
	send_end = nullptr;
	write_buff_end = &buffer[buffer_len - 1];
	memset(buffer, 0, write_end - write_position);
}
//////////////////////////////////////////////////////////////
//write data
void NetworkStream::BeginWrite()
{
	write_position = write_end;
	write_end += 4;
}
void NetworkStream::EndWrite()
{
	int totole_len = write_end - write_position;
	int int_len = sizeof(int);
	if (totole_len > int_len)
	{
		int data_len = totole_len - int_len;
		char* write_end_backup = write_end;
		write_end = write_position;
		WriteInt(data_len);
		write_end = write_end_backup;
		char* start = send_position ? send_position : write_position;
		int wait_send_len = write_end - start;
		memmove(write_buff, start, wait_send_len);
		send_position = write_buff;
		send_end = send_position + wait_send_len;
		write_position = write_end = send_end;
	}
	else
	{
		write_end = write_position;
	}
}
void NetworkStream::WirteByte(byte data)
{
	WriteData(&data, sizeof(byte));
}
void NetworkStream::WirteByte(char data) 
{
	WriteData(&data, sizeof(char));
}
void NetworkStream::WriteShort(short data) 
{
	WriteData(&data, sizeof(short));
}
void NetworkStream::WriteUShort(ushort data) 
{
	WriteData(&data, sizeof(ushort));
}
void NetworkStream::WriteInt(int data) 
{
	WriteData(&data, sizeof(int));
}
void NetworkStream::WriteUInt(uint data)
{
	WriteData(&data, sizeof(uint));
}
void NetworkStream::WriteFloat(float data) 
{
	WriteData(&data, sizeof(float));
}
void NetworkStream::WriteLong(long data) 
{
	WriteData(&data, sizeof(long));
}
void NetworkStream::WriteULong(ulong data) 
{
	WriteData(&data, sizeof(ulong));
}
void NetworkStream::WriteString(const char* str) 
{
	int len = strlen(str);
	WriteInt(len);
	WriteData(str, len);
}
void NetworkStream::WriteData(const void* data, int count) 
{
	int empty_size = (write_buff + write_buff_len) - (write_end + count);
	if (write_buff == nullptr || count < 0 || empty_size <= 0)
	{
		OnStreamError(WRITEERROR);
		return;
	}
	if (count > 0)
	{
		memcpy(write_end, data, count);
		write_end += count;
	}

}
bool NetworkStream::WriteProto(ProtoMessage &proto)
{
	int size = proto.ByteSize();
	if (size > 0 && ((write_buff_end - write_position) - (size + sizeof(int)))>0)
	{
		WriteInt(size);
		proto.SerializeToArray(write_end, size);
		write_end += size;
		return true;
	}
	return false;
}
//////////////////////////////////////////////////////////////
//read data
void NetworkStream::ReadByte(byte &data)
{
	ReadData(&data, sizeof(byte));
}
void NetworkStream::ReadByte(char &data)
{
	ReadData(&data, sizeof(char));
}
void NetworkStream::ReadShort(short &data)
{
	ReadData(&data, sizeof(short));
}
void NetworkStream::ReadUShort(ushort &data)
{
	ReadData(&data, sizeof(ushort));
}
void NetworkStream::ReadInt(int &data)
{
	ReadData(&data, sizeof(int));
}
void NetworkStream::ReadUInt(uint &data)
{
	ReadData(&data, sizeof(uint));
}
void NetworkStream::ReadFloat(float &data)
{
	ReadData(&data, sizeof(float));
}
void NetworkStream::ReadLong(long &data)
{
	ReadData(&data, sizeof(long));
}
void NetworkStream::ReadULong(ulong &data)
{
	ReadData(&data, sizeof(ulong));
}
int NetworkStream::ReadString(char* str,int size)
{
	int len = 0;
	ReadInt(len);
	if (len < 0||size<len)
	{
		throw READERROR;
		
	}
	str[len] = 0;
	ReadData(str, len);
	return len;
}
void NetworkStream::ReadData(void* data, int count)
{
	if (read_buff == nullptr || count < 0 || read_end - read_position<count)
	{
		OnStreamError(READERROR);
		return;
	}
	if (count > 0)
	{
		memcpy(data, read_position, count);
		read_position += count;
	}
}
bool NetworkStream::ReadProto(ProtoMessage &proto)
{
	int size = 0;
	ReadInt(size);
	int unread = read_end - read_position;
	if (size > 0&& unread>=size)
	{
		proto.ParseFromArray(read_position, size);
		read_position += size;
		return true;
	}
	else
	{
		OnStreamError(READERROR);
	}
	return false;
}